### ekf config file ###
ekf_filter_node:
    ros__parameters:
        publish_tf: true
        map_frame: map              # Defaults to "map" if unspecified
        odom_frame: odom            # Defaults to "odom" if unspecified
        base_link_frame: base_link  # Defaults to "base_link" if unspecified
        world_frame: odom            # Defaults to the value of odom_frame if unspecified

        frequency: 10.0
        two_d_mode: true
        sensor_timeout: 0.01
        
        # -------------------------------------
        # External IMU:
        # -------------------------------------
        imu0: imu/test
        # make sure the input is aligned with ROS REP105. "imu_correct" is manually transformed by myself. EKF can also transform the data using tf between your imu and base_link
        imu0_config: [false, false, false,
                        true, true, true,
                        false, false, false,
                        true, true, true,
                        false, false, false]
        imu0_differential: false
        imu0_relative: false
        imu0_queue_size: 10 
        imu0_remove_gravitational_acceleration: true

        # gps0: "/ublox_gps_node/fix"
        # gps0_config: [true,  true,  false,  # X, Y, Z 위치
        #             false, false, false,  # X, Y, Z 속도
        #             false, false, false,  # X, Y, Z 가속도
        #             false, false, false]  # X, Y, Z 방향
        # gps0_differential: false
        # gps0_queue_size: 10

        # # -------------------------------------
        # # Odometry (From Navsat):
        # # -------------------------------------

        odom0: odometry/gps
        odom0_config: [true, true, false,
                        false, false, false,
                        false, false, false,
                        false, false, false,
                        false, false, false]
        odom0_differential: false
        odom0_queue_size: 10
        odom0_relative: true

        # odom1: odometry/wheel
        # odom1_config: [true, true, false,
        #             false, false, false,
        #             true,  true,  false,
        #             false, false, true,
        #             false, false, false]
        # odom1_queue_size: 10
        # odom1_differential: false
        # odom1_relative: false

        # twist0: erp42/twist
        # twist0_config: [false, false, false,
        #                 false, false, false,
        #                 true,  false,  false,
        #                 false, false, false,
        #                 false, false, false]
        # twist0_differential: false                
        # twist0_queue_size: 10
        # twist_relative: false

        # pose0: pcl_pose
        # pose0_config: [true, true, false,
        #                 false, false, true,
        #                 false, false, false,
        #                 false, false, false,
        #                 false, false, false]
        # pose0_differential: false
        # pose0_queue_size: 10
        # pose0_relative: false


        # imu0: imu/data
        # # make sure the input is aligned with ROS REP105. "imu_correct" is manually transformed by myself. EKF can also transform the data using tf between your imu and base_link
        # imu0_config: [false, false, false,
        #                 false, false, false,
        #                 false, false, false,
        #                 false, false, true,
        #                 true, true, false]
        # imu0_differential: false
        # imu0_queue_size: 50 
        # imu0_relative: false
        # imu0_remove_gravitational_acceleration: true


        process_noise_covariance: [1.0, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    1.0, 0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    10.0, 0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.03, 0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.03, 0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.1, 0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.25, 0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.25, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.04, 0.0,    0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.01, 0.0,    0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.01, 0.0,    0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.05, 0.0,    0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.01, 0.0,    0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.01, 0.0,
                                        0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.015]

        